### ISSUE TITLE
Custom sorting/grouping of options in the generated --help messages

### ISSUE THREAD

**Issue by squeakyneb** [scallop/scallop#196]
I'm writing a tool at the moment with many options. It would be helpful to be able to divide the --help message into sections (e.g. basic options and advanced options, options specific to a particular capability of the tool that would otherwise be alphabetically sorted and scattered through the --help listing)

The only way I can see to do this currently is with subcommands, which changes the syntax of the command in non-meaningful way. I suppose I could also add a prefix to options to group them, but this also seems a bit silly as I'm making the user type extraneous characters so that the help message can be better, which seems like a silly and orthogonal trade-off.

I think it would be ideal to have some sort of group directive that would have a name and produce a help output similar to how subcommands work. This could perhaps also be done as a string or symbol as an extra optional parameter on opt/ScallopOption.

**Comment by Rogach** [comment]
You can currently do something like this with `mainOptions` setting - it will place several options before all others in defined order. For example:

```scala
object Conf(args: Seq[String]) extends ScallopConf(args) {
  val alpha = opt[Int]()
  val beta = opt[Int]()
  val gamma = opt[Int]()
  val delta = opt[Int]()
  
  mainOptions = Seq(beta, delta)

  verify()
}
```

**Comment by squeakyneb** [comment]
Yeah I was having a look at that. I still think more granular divisions and named divisions would be a useful feature though.

(also I think you mean `mainOptions = Seq(...)`,for the sake of anyone else reading this in the future getting confused 🙂 )

**Comment by Rogach** [comment]
I'll think about what can be done. I'm heavily short on time at the moment, though - won't be able to implement it quickly.

> also I think you mean mainOptions = Seq(...)

Yes, you are right - I got confused in a hurry.

**Comment by squeakyneb** [comment]
Yeah that's understandable. I'm otherwise very much appreciating your work.

I'd take a shot at it myself but I'm not sure how the scala-native and scala-js stuff works 🤔

**Comment by Rogach** [comment]
Regarding scala-native and scala-js - they work almost without issues. I kept 99% of the code intact when adding support for them. The only issue was with reflection that was used for option name guessing - reflection is obviously not available in native and js targets.

**Comment by GuyFleegman** [comment]
I would like this feature too.
Right now, my app has several distinct groups of options: one for setting the verbosity level (--debug, --trace, --warn, etc.), one for the main commands (--search, --count, --list, etc.), one for additional options.... As it is, these are all intermixed in the --help output.  :-P
Also - I wasn't able to add my propsLong option to the mainOptions.  That especially sucks since searching (--search name=value) is my app's main use case.

**Comment by Rogach** [comment]
I propose the following syntax:

```scala
object Conf extends ScallopConf {
  private val mainGroup = group()
  val search = opt[String](group = mainGroup)
  val count = opt[String](group = mainGroup)
  val list = opt[String](group = mainGroup)

  private val verbosityLevels = group(header = "Verbosity levels:")
  val debug = opt[Boolean](group = verbosityLevels)
  val trace = opt[Boolean](group = verbosityLevels)
  val warn = opt[Boolean](group = verbosityLevels)
  
  // Groups are output in help in order of group() invocation appearance in config.
  // mainOptions that we used previously are kept for now (and shown before any other options/groups),
  // but they are deprecated in favor of this new mechanism.
  // All options without `group` parameter are sorted and placed after all the explicit groups.
}
```

What do you think?

**Comment by Rogach** [comment]
Released in version `4.0.0`.
